sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+  thrown even if the aspect does not correspond to any optional feature.
+
+* For a kernel that is decorated with `[[sycl::requires()]]`: the exception
+  must be thrown if a function in the kernel's static call tree uses a feature
+  that the device does not support even if the `[[sycl::requires()]]` attribute
+  is missing the corresponding aspect.  This case can only occur if the kernel
+  calls a `SYCL_EXTERNAL` function in another translation unit and the function
+  uses an optional feature that is not listed in the `[[sycl::requires()]]`
+  attribute attached to the `SYCL_EXTERNAL` function declaration.  (In any
+  other case, the front-end compiler would diagnose an error.)  Although the
+  SYCL 2020 spec says that such applications are non-conformant, it is easy for
+  DPC++ to throw an exception in such a case.
+
+* The kernel is decorated with the `[[sycl::reqd_work_group_size(W)]]` or
+  `[[sycl::reqd_sub_group_size(S)]]` attribute, and the device does not support
+  the work group size `W` or the sub-group size `S`.
+
+Note that the exception must be thrown synchronously, not delayed and thrown on
+the queue's asynchronous handler.
+
+### No runtime exception for speculative compilation
+
+It is currently common for the runtime to speculatively compile some kernels.
+For example, DPC++ may bundle all kernels from the same translation unit
+together into a single device image.  When the application submits one kernel K
+to a device D, the runtime actually compiles all kernels in K's device image
+for device D.  Let's assume in this example that the kernel K uses only
+features that are supported by D.  It would be illegal for the runtime to throw
+an exception in such a case just because some other kernel in the same device
+image uses a feature that is not supported by device D.
+
+
+## Design to implement required diagnostics
+
+### Changes to DPC++ headers
+
+With the exception of the `[[sycl::reqd_work_group_size()]]` and
+`[[sycl::reqd_sub_group_size()]]` attributes, all kernel optional features
+are associated with some device aspect.  For example, the `sycl::half` type
+is an optional feature which is only supported on devices that have the
+`aspect::fp16` aspect.  We can therefore use device aspects as a way to
+describe the set of optional features that a kernel uses (with the exception
+of the required work-group or sub-group size).
+
+As we will see later, it will be very helpful to decorate all APIs in DPC++
+headers that correspond to optional kernel features with the
+`[[sycl::requires()]]` attribute.  For example, the declaration of the
+`sycl::half` type would look like this:
+
+```
+using half [[sycl::requires(has(aspect::fp16))]] = cl::sycl::detail::half_impl::half;
+```
+
+In cases where the optional feature corresponds to use of a class (e.g.
+`sycl::atomic_ref`), the declaration can look like this:
+
+```
+template</*...*/>
+class [[sycl::requires(has(aspect::atomic64))]] atomic_ref {
```

Comment:
I saw your previous comment.  My response might be hidden because I marked the conversation "resolved".  I agree with what you say, but I don't want to put a long code snippet at this point in the document because it will take attention away from the main point of this section.  Instead, I added an example of the partial specialization that will be needed in an appendix at the end of the document.  I also added a note just a few line below here that points to that appendix.

Reply:
This example then should better be replaced or removed, in my opinion, but I'll leave this to you to decide.

Reply:
I replaced the example with a simpler one, but kept the appendix showing how `atomic_ref` can be specialized to only use the attribute for an 8-byte type.